lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,195 @@
+      def client
+        config = OCI::Config.new
+        config.user = secrets['user']
+        config.fingerprint = secrets['fingerprint']
+        config.key_content = secrets['key_content']
+        config.tenancy = secrets['tenancy']
+        config.region = secrets['region']
+        config.validate
+        OCI::Dns::DnsClient.new(config: config)
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        client.get_zone_records(zone)
+          .flat_map { |response| response.data.items }
+          .flat_map { |api_record| build_from_api(api_record) }
+          .compact
+      rescue StandardError
+        stdout.puts "Cannot build recrod for zone: #{zone}"
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets['compartment_id']).data.map(&:name)
+      end
+
+      private
+
+      # Returns the version of the zone
+      def zone_version(zone)
+        client.get_zone(zone).data.version
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record_rdata,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_add_record)
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = Record.ensure_ends_without_dot(record.fqdn)
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+        found_record = client.get_zone_records(
+          zone,
+          rtype: record.type,
+          zone_version: client.get_zone(zone).data.version,
+          domain: record_fqdn,
+        ).data.items.last
+
+        begin
+          record_hash = found_record.record_hash
+          patch_remove_record = [
+            OCI::Dns::Models::RecordOperation.new(
+              domain: record_fqdn,
+              record_hash: record_hash,
+              rtype: record.type,
+              ttl: record.ttl,
+              rdata: record_rdata,
+              operation: 'REMOVE',
+            ),
+          ]
+
+          client.patch_zone_records(
+            zone,
+            OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_remove_record)
+          )
+        rescue NoMethodError
```

Comment:
Feels like a `NoMethodError` should be caught before actually calling the method on something that does not have the method ? What creates this ?

Reply:
In case of trying to remove a record that doesn't exist, 
`found_record` will be `nil` and `found_record.record_hash` will be calling `NoMethodError`.
Not sure this'd be good way to catch it, just followed the sequence to catch the error.

Reply:
You probably would want a:

```ruby
raise SomeError, "record not found" unless found_record
```

before calling the `record_hash` method then!

Reply:
However it's implemented, I think it's important that the behaviour is well defined (write a test so changing the behaviour here may only be done deliberately).

@XaF : For what it's worth, I agree that making it idempotent makes sense:

1) It seems like the NS1 provider ends up doing this:  https://github.com/Shopify/record_store/blob/956cb98c0cd035acb128683cb1d83e1dd4bf5a53/lib/record_store/provider/ns1.rb#L78-L107

2) In this case, at least, deleting a record has no side effects besides the desired post-condition (the target record no longer exists).  So just ensuring that post-condition holds feels like all we'd need to do...if it already holds before we do anything, well, ... great!  If deleting an already deleted record had side effects (e.g. if the extraneous work put pressure on a rate limit, or it caused some other problem) complaining could possibly start making sense.  With what we currently know, there's no reason to complain.

tl;dr - @gooallen choose an option (raise, or don't raise), but write a test describing what your desired behaviour is.  Not raising if the record doesn't exists gets my vote as well.

Reply:
Cool, I wrote a test for it and I will go with this guy.
```
return unless found_record  # No need to do the following as the record does not exist
record_hash = found_record.record_hash
```
I was also thinking this is good option if `round_record` is nil, no patch with record_store then won't cause any errors.